### ISSUE TITLE
JIT: consider running some optimizations in Tier0

### ISSUE THREAD

**Issue by AndyAyersMS** [dotnet/coreclr#14474]
The goal of Tier0 is to jit code as quickly as possible. Currently this is done by having Tier0 enable the "minopts" mode in the jit, which disables all optimization. But there are good reasons to believe that running some optimizations in Tier0 can improve the speed of jitting.

To first order the time it takes to jit a method (especially when lightly optimizing) is proportional to the amount of code the jit produces for the method. So any cheap optimization that reduces the overall size of the generated code is a candidate for running in Tier0. Some ideas:
* enable the importer branch folding
* enable the early type opts that feed importer branch folding. Perhaps doubly relevant since R2R prejitting will leave generic code to the jit, so tier0 will see more generic instantations then "normal", and these are the method bodies that often can be greatly simplified by early type opts. These opts are also pretty cheap.
* other kinds of simple expression tree simplifications -- maybe some parts of morph are cheap enough to enable?
* avoid inline expansion of helpers. Tricky because while this helps jit time it slows down the jitted code -- but we hope not to run the Tier0 code very often, so it seems like it could pay off.
* perhaps inline very small methods. I have some older and perhaps flawed data that indicates this should be TP win, but recent results haven't borne this out. But it is worth revisiting. Certainly, inlining small methods often reduces code size, and it should cut down on the number of jit invocations.

There are notes about this scattered about in other issues, Will track these down and link them back here.

**Comment by AndyAyersMS** [comment]
Also:
* #13589 use more compact IR representation for array access
* #13280 change approach in the jit to maintaining local var ref counts

**Comment by AndyAyersMS** [comment]
Considering for 2.1.

**Comment by AndyAyersMS** [comment]
Did some prototyping of this a while back and wasn't really able to get a clear picture of potential improvements. So am going to hold off on this until after 2.1.

**Comment by tannergooding** [comment]
@AndyAyersMS, is removing dead code for constant branches currently part of this?

There are a few places today where we return `gtIconNode` (such as Hardware Intrinsics, SIMD Intrinsics, etc) that look like they are clear/easy wins for minopts.

**Comment by AndyAyersMS** [comment]
Yeah that's the "importer branch folding" bit mentioned above.

I agree there are be wins to be had here. At the time I ran these investigations, I was not seeing consistent wins on the scenarios I was able to measure. But it was early and the measurements were somewhat ad-hoc. I believe @noahfalk is working on a more comprehensive set of well defined scenarios to use in evaluations. Would be nice to have something in there that leverages intrinsics.

So I'll revisit this, but likely not until after 2.1 is more or less done.

**Comment by AndyAyersMS** [comment]
See #22984 for some notes on a prototype: [EnableBoxingOptsTier0](https://github.com/dotnet/coreclr/compare/master...AndyAyersMS:EnableBoxingOptsTier0).

**Comment by AndyAyersMS** [comment]
We should also consider enabling intrinsic expansion, for both HW intrinsics and for intrinsics that are likely to lead to early control flow trimming.
